README.md
![alt text align:center](https://www.sbaum.com.br/images/logo-mini.png "Sbaum")

# NFeCloud - SDK PHP


## Descrição
Este pacote consiste em um SDK em PHP para a utilizacao do sistema NFeCLoud através de api - REST.

# Requisitos
- PHP >=5.5.19;


# Instalação

Via Composer

```bash
composer require sbaum-tecnologia/nfecloud-php
```

## Licença
GNU GPLv3. Por favor, veja o [Arquivo de Licença](license.txt) para mais informações.

## Exemplo
```php
require __DIR__.'/../vendor/autoload.php';

// Coloca o Token NFECLOUD  no environment do PHP.

putenv('NFECLOUD_TOKEN=AAAA');
putenv('NFECLOUD_TOKEN_SECRET=aBBBCC');


// Cria Objetos de Consulta/Servico
 

$ServiceEmpresas = new NFeCloud\Empresas();
$ServiceNotas = new NFeCloud\Notas();
$ServiceActions = new NFeCloud\Actions();


// Obtem Todas as Empresas vinculadas a conta 
// 'order' => ordena pelo atributo especificadp [asc=>ascendente, desc=>descendente]
// 'limit' => qual o numero de elementos deve buscas (máximo 100)
// 'page' => qual a página ( para mostrar do elemento 201 ao 300 busque pela página 3)
// 'filtros' => array de condicões para pesquisa no formato: array('atributo',criteria('=','>=','<=','>','<','like,'IN'),'valor')
// 
//  retorno: array com ids encontrados
 //

try {        
    
    $ret = $ServiceEmpresas->all(['order'=>'cnpj desc','limit'=>10,'page'=>1,'filtros'=>[['id','=','107']]]);    
    foreach ($ret->ids as $id){
        
         // Busca a empresa pelo id         
        $_ret = $ServiceEmpresas->get($id);
        echo "CNPJ: " . $_ret->cnpj . "</br>";
        echo "ID: " . $_ret->id . "</br>";
        
        // busca as notas que pertence a empresa
        // Para buscas pelo nsu - numero sequencial para cada emissor gerado pela Sefaz - formatar conforme exemplo
        
        $value = 7200;
        $nsu = str_pad($value, 15, '0', STR_PAD_LEFT);        
        
        
        $retN = $ServiceNotas->all(['order'=>'dhEmi desc','limit'=>2,'page'=>1,'filtros'=>[['empresas_id','=',$_ret->id],['nsu','>=',$nsu],['xml_arquivado','=',1]]]);        
        foreach ($retN->ids as $id){
                
                // Busca a nota pelo id
                // o atributo xml retorna em base64
                $_retN = $ServiceNotas->get($id);
                echo "----</br>";
                echo "ID: " . $_retN->id . "</br>";
                echo "Numero: " . $_retN->numero . "</br>";
                echo "Emissao: " . $_retN->dhEmi . "</br>";
                echo "Emitente/Destinatario: " . $_retN->xNome . "</br>";
                //echo "xml: " . base64_decode($_retN->xml) . "</br>";
                
                // Consulta Status da NF-e - parametro id da nota
                 
                $ret_Consulta = $ServiceActions->consultarStatusNFE((int)$_retN->id) ;
                echo $ret_Consulta . "</br>";
                
                //Faz a manifestação da NF-e - parametro id da nota, tipo de manifestação, jusitificativa
                
                $ret_Consulta =  $ServiceActions->manifestarNFE((int)$_retN->id, '210210', '');
                echo $ret_Consulta . "</br>";
                echo "----</br>";
                
                 //Faz download do xml e/ou pdf - parametro id da nota, xml (true/false), pdf (true/false)
                 
                $ret_Consulta =  $ServiceActions->downloadXMLPDF((int)$_retN->id, true,true);                
                print_r($ret_Consulta) . "</br>";
                echo "----</br>";
         }
    }
    
    
    // Consulta novos documentos (DFE) emitidas para a empresa
    //DFE - Resumos de Notas // XML - arquivos xml // Eventos - Eventos vinculados 
    
    $ret_Consulta =  $ServiceActions->consultarDFeSefaz((int)'107');
    print_r($ret_Consulta);
    
    //Upload de arquivos xml (NF-e/NFC-e/CT-e/NFS-e/Eventos)
         
    $conteudo = file_get_contents(__DIR__.'/41170107464573000115550010000060971776089318nfe.xml');
    $xml = base64_encode($conteudo);
    $ret_upl = $ServiceActions->uploadXML($xml);
    echo $ret_upl;
    echo '--Fim--';
} catch (Exception $e) {
    echo $e->getMessage();    
}

CHANGELOG.md
## 1.0.0 - 03/01/2017
- Initial version.


src/Actions.php
<?php

namespace NFeCloud;

class Actions extends Resource
{
        
    /**
     * The endpoint that will hit the API.
     *
     * @return string
     */
    public function endpoint()
    {
        return 'do';
    }
    
    public function all(array $params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    
    }
    
    public function create(array $form_params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    }
    
    public function retrieve($id = null)
    {
        throw new \Exception($this->msg_NoAvaliable);
    }
    
    public function update($id = null, array $form_params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    }
    
    public function delete($id = null, array $form_params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    }
    
    public function get($id = null, $additionalEndpoint = null,array $params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    }
    
    public function post($id = null, $additionalEndpoint = null, array $form_params = [])
    {
        throw new \Exception($this->msg_NoAvaliable);
    }

   
    
    /**
     * Faz busca de DFE (Documento FIscal Eletronico) junto a Sefaz.
     *
     * @param int    $id                  id da Empresa a ser realizada a busca.     *
     
     * @return mixed
     */
    public function consultarDFeSefaz($id){
        if(!is_int($id)){
            throw new \Exception('Informe a id da empresa');
        }        
        return parent::post(null,'consultarDFeSefaz',['empresas_id'=>$id]);
    }
    
    /**
     * Faz a consulta do status da NF-e/NFC-e junto a SEFAZ.
     *
     * @param int    $id                 id da Nota a ser Consultada.     *
     *
     * @return string
     */    
    public function consultarStatusNFE($id){
        if(!is_int($id)){
            throw new \Exception('Informe a id da nota para consulta status');
        }
        return parent::post(null,'consultarStatusNFE',['id'=>$id]);
    }
    
    /**
     * Faz a manifestacao da NF-e junto a SEFAZ.
     *
     * @param int    $id                  id da Nota a ser manifestada.     *
     * @param string $notas_manifesto_id  id do manifesto 
     *                                    210200 – Confirmação da Operação
     *                                    210210 – Ciência da Operação
     *                                    210220 – Desconhecimento da Operação
     *                                    210240 – Operação não Realizada   
     * @param string $txtJustificativa    (15-255) justificativa do manifesto Obrigatória para o tipo 210240 – Operação não Realizada  
     * @return string
     */    
    public function manifestarNFE($id,$notas_manifesto_id,$txtJustificativa){
        if(!is_int($id)){
            throw new \Exception('Informe a id da nota para manifestar');
        }
        if(!in_array($notas_manifesto_id, ['210200','210210','210220','210240']) ){
            throw new \Exception('Informe o codigo do tipo de Evento - (210200,210210,210220,210240)');
        }
        if($notas_manifesto_id == '210240' && (strlen($txtJustificativa) < 15 ||  strlen($txtJustificativa) > 255)){
            throw new \Exception('É obrigatória uma justificativa com 15 até 255 caracteres!! Evento - 210240');
        }
        return parent::post(null,'manifestarNFE',['id'=>$id,'notas_manifesto_id'=>$notas_manifesto_id,'txtJustificativa'=>$txtJustificativa]);
    }
    
    
    /**
     * Upload de Documentos XML - Empresas (Emitente ou Destinatário Pertencente a Conta).
     *     
     * @param string $xml   Conteudo do Arquivo XML - Base64
     
     * @return string
     */
    public function uploadXML($xml){
        if(strlen($xml)==0){
            throw new \Exception('Informe o Conteudo do Arquivo - base64');
        }
        return parent::post(null,'uploadXML',['xml'=>$xml]);
    }
    
    
    public function downloadXMLPDF($id,$xml=true,$pdf=true){
        if(!is_int($id)){
            throw new \Exception('Informe a id da nota');
        }
        $data = array();
        $data['id'] = $id;
        $data['xml'] = '0';
        $data['pdf'] = '0';
        if($xml){
            $data['xml'] = '1';
        }
        if($pdf){
            $data['pdf'] = '1';
        }
        return parent::post(null,'downloadXMLPDF',$data);        
    }
    
    
}
